### PR TITLE
fix(form): resolve the problem of validation states changing out of sequence

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [X] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- https://github.com/ant-design/ant-design/issues/41384

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | resolve the problem of validation states changing out of sequence |
| 🇨🇳 Chinese | 解决验证状态不按照顺序改变的问题 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

### CODE REVIEW

File: components/form/FormItem/ItemHolder.tsx [hunk 1]
```diff
@@ -86,24 +86,49 @@ export default function ItemHolder(props: ItemHolderProps) {
...
     // success feedback should display when pass hasFeedback prop and current value is valid value
     mergedValidateStatus = 'success';
   }
 
+  const desplayValidateStatus = React.useMemo(() => {
+    let status: ValidateStatus = '';
+    if (validateStatus !== undefined) {
+      status = validateStatus;
+    } else if (meta.validating) {
+      status = 'validating';
```
Comment (yoyo837) [inline components/form/FormItem/ItemHolder.tsx:103]: 这么多 else if, 改为 if return吧

Comment (kiner-tang) [reply components/form/FormItem/ItemHolder.tsx]: done

File: components/form/FormItem/ItemHolder.tsx [hunk 2]
```diff
@@ -86,24 +86,40 @@ export default function ItemHolder(props: ItemHolderProps) {
...
-  } else if (debounceWarnings.length) {
+  } else if (meta.warnings.length) {
     mergedValidateStatus = 'warning';
   } else if (meta.touched || (hasFeedback && meta.validated)) {
     // success feedback should display when pass hasFeedback prop and current value is valid value
     mergedValidateStatus = 'success';
   }
 
+  const desplayValidateStatus = React.useMemo(() => {
+    if (validateStatus !== undefined) return validateStatus;
```
Comment (yoyo837) [inline components/form/FormItem/ItemHolder.tsx:99]: 个人觉得保留括号好一些

Comment (afc163) [reply components/form/FormItem/ItemHolder.tsx:99]: +1

Comment (kiner-tang) [reply components/form/FormItem/ItemHolder.tsx]: ok

File: components/form/FormItem/ItemHolder.tsx [hunk 3]
```diff
@@ -86,24 +86,50 @@ export default function ItemHolder(props: ItemHolderProps) {
...
     mergedValidateStatus = 'error';
-  } else if (debounceWarnings.length) {
+  } else if (meta.warnings.length) {
     mergedValidateStatus = 'warning';
   } else if (meta.touched || (hasFeedback && meta.validated)) {
     // success feedback should display when pass hasFeedback prop and current value is valid value
     mergedValidateStatus = 'success';
   }
 
+  const desplayValidateStatus = React.useMemo(() => {
```
Comment (zombieJ) [inline components/form/FormItem/ItemHolder.tsx:98]: 差不多意思，但是逻辑有一点点重复，和上面一起抽一个方法出来？

Comment (kiner-tang) [reply components/form/FormItem/ItemHolder.tsx]: done

File: components/form/__tests__/index.test.tsx [hunk 4]
```diff
@@ -1626,4 +1626,51 @@ describe('Form', () => {
...
+  it('validate status should be change in order', async () => {
+    const onChange = jest.fn();
+
+    const CustomInput = (props: any) => {
+      const { status } = Form.Item.useStatus();
+      useEffect(() => {
+        onChange(status);
+      }, [status]);
+      return <Input {...props} />;
+    };
```
Comment (miracles1919) [inline components/form/__tests__/index.test.tsx:1638]: 这里是不是可以不用 useEffect，不然还要区分 react 版本
```tsx
const CustomInput = (props: any) => {
    const { status } = Form.Item.useStatus();
    return <><Input {...props} />{status}</>;
};

changeValue(0, '1');
expect(status).toBe('validating')
expect(status).toBe('error')
```

Comment (kiner-tang) [reply components/form/__tests__/index.test.tsx:1638]: 主要是为了模拟用户的使用场景进行测试

**Review by zombieJ — APPROVED** [review]

**Pull request merged**